system/lib/pthread/proxying.c

```diff
@@ -31,45 +32,49 @@
-// processed, we decrease the reference count. The queue can only be freed once
-// `em_proxying_queue_destroy` has been called and the reference count has
-// reached zero.
+// notification containing a reference to the a thread-local task queue is
+// generated, we set a flag on that task queue. Each time that task queue is
+// processed, we clear the flag. The proxying queue can only be freed once
+// `em_proxying_queue_destroy` has been called and the notification flags on
+// each of its task queues have been cleared.
 //
 // But an extra complication is that the target thread may have died by the time
 // it gets back to its event loop to process its notifications. This can happen
 // when a user proxies some work to a thread, then calls
 // `emscripten_proxy_execute_queue` on that thread, then destroys the queue and
 // exits the thread. In that situation no work will be dropped, but the thread's
-// worker will still receive a notification and have to decrease the reference
-// count without a live runtime. Without a live runtime, there is no stack, so
+// worker will still receive a notification and have to clear the notification
+// flag without a live runtime. Without a live runtime, there is no stack, so
 // the worker cannot safely free the queue at this point even if the refcount
 // goes to zero. We need a separate thread with a live runtime to perform the
 // free.
 //
 // To ensure that queues are eventually freed, we place destroyed queues in a
-// global "zombie list" where they wait for their refcounts to fall to zero. The
-// zombie list is scanned whenever a new queue is constructed and any of the
-// zombie queues with zero refcounts are freed. In principle the zombie list
-// could be scanned at any time, but the queue constructor is a nice place to do
-// it because scanning there is sufficient to keep the number of zombie queues
-// from growing without bound; creating a new zombie ultimately requires
-// creating a new queue.
-
-extern int _emscripten_notify_proxying_queue(pthread_t target_thread,
-                                             pthread_t curr_thread,
-                                             pthread_t main_thread,
-                                             em_proxying_queue* queue);
+// global "zombie list" where they wait for their notification flags to be
+// cleared. The zombie list is scanned whenever a new queue is constructed and
+// any of the zombie queues without outstanding notifications are freed. In
+// principle the zombie list could be scanned at any time, but the queue
+// constructor is a nice place to do it because scanning there is sufficient to
+// keep the number of zombie queues from growing without bound; creating a new
+// zombie ultimately requires creating a new queue.
 
 typedef struct task {
   void (*func)(void*);
   void* arg;
 } task;
 
+typedef enum notification_state {
+  NO_NOTIFICATION,
+  RECEIVED_NOTIFICATION,
+  PENDING_NOTIFICATION,
+} notification_state;
```

Comment:
Please add the constants here (` = 0` etc.) as they are used from JS too, so it's nice to not need to figure them out when reading the JS.

Also please add a comment on each state. Reading this far, I'm not sure what they mean yet.